In [4]:
import nltk
import re

In [1]:
'hello world'

a=''' Basic Regular Expression Meta-Characters, Including Wildcards, Ranges and Closures
.	        Wildcard, matches any character
^abc	    Matches some pattern abc at the start of a string
abc$	    Matches some pattern abc at the end of a string
[abc]	    Matches one of a set of characters
[^abc]      Matches anything but a set of characters
[A-Z0-9]	Matches one of a range of characters
ed|ing|s	Matches one of the specified strings (disjunction)
*	        Zero or more of previous item, e.g. a*, [a-z]* (also known as Kleene Closure)
+	        One or more of previous item, e.g. a+, [a-z]+
?	        Zero or one of the previous item (i.e. optional), e.g. a?, [a-z]?
{n}	        Exactly n repeats where n is a non-negative integer
{n,}	    At least n repeats
{,n}	    No more than n repeats
{m,n}	    At least m and no more than n repeats
a(b|c)+	    Parentheses that indicate the scope of the operators

SyntaxError: EOF while scanning triple-quoted string literal (<ipython-input-1-5f8c27f9b80b>, line 17)

* wordlist의 타입 = list
# * regular expression *

* list가 아니라 string에만 적용된다.
* ( ) 안에 들어 있는 것이 입력

In [9]:
re.search('abc', 'abced')

<_sre.SRE_Match object; span=(0, 3), match='abc'>

* 입력이 2개인 함수이다.
* abcde라는 것에서 abc를 찾아라.

In [32]:
wordlist = [w for w in nltk.corpus.words.words('en') if w.islower()]
print(wordlist[:10])

['a', 'aa', 'aal', 'aalii', 'aam', 'aardvark', 'aardwolf', 'aba', 'abac', 'abaca']


In [37]:
result = [w for w in wordlist if re.search('ed$',w)][:10]
print(result)

['abaissed', 'abandoned', 'abased', 'abashed', 'abatised', 'abed', 'aborted', 'abridged', 'abscessed', 'absconded']


* wordlist는 string들의 list
* re.search는 list가 아니라 string에만 적용되기 때문에 for루프를 돌려서 하나씩 넣어주어야 한다.
* 'ed$' 는 ed로 끝나는 것을 찾으라는 것, [:10]은 앞에서부터 10개를 표시하라는 것. 

In [39]:
result = [w for w in wordlist if re.search('^..j..t..$' , w)][:10]
print(result)

['abjectly', 'adjuster', 'dejected', 'dejectly', 'injector', 'majestic', 'objectee', 'objector', 'rejecter', 'rejector']


* . 은 하나의 캐릭터다. 3번째가 j, 6번째가 t인 8글자의 단어를 앞에서부터 10개 찾아라.

In [41]:
result = [w for w in wordlist if re.search('^[ghi][mno][jlk][def]$', w)]
print(result[:10])

['gold', 'golf', 'hold', 'hole']


* []으로 묶으면 그 중 무엇이든 상관 없다는 뜻
* matches one of a set of characters. 
* ex) hold, gold, golf, hole

In [13]:
wsj = sorted(set(nltk.corpus.treebank.words()))

In [28]:
wordlist = [w for w in wsj if re.search('^[0-9]+\.[0-9]+$', w)]
print(wordlist[:10])

['0.0085', '0.05', '0.1', '0.16', '0.2', '0.25', '0.28', '0.3', '0.4', '0.5']


* +는 one or more
* ex) a, aa, ah, aha, h, ha, hah

In [27]:
wordlist = [w for w in wsj if re.search('^[0-9]+\.[0-9]$', w)]
print(wordlist[:10])

['0.1', '0.2', '0.3', '0.4', '0.5', '0.7', '0.9', '1.1', '1.2', '1.4']


* [0-9]는 0에서 9라는 뜻이고, +는 숫자가 반복되어도 된다는 의미이다. 따라서 어떤 숫자도 표시 가능.


In [43]:
word = 'supercalifragilisticexpialidocious'

In [47]:
result = re.findall(r'[aeiou]', word)
print(result)

['u', 'e', 'a', 'i', 'a', 'i', 'i', 'i', 'e', 'i', 'a', 'i', 'o', 'i', 'o', 'u']


* 이 word, supercalifragilisticexpialidocious에서 모음을 찾아라

In [50]:
result = re.findall('[aeiou](..)[aeiou]', word)
print(result)

['rc', 'fr', 'st', 'xp', 'ci']


* 모음 + 2글자 + 모음을 찾아라

In [48]:
re.findall('ab' , 'ababdddabkabddd')

['ab', 'ab', 'ab', 'ab']

In [49]:
fdist = nltk.FreqDist(vs for word in wsj for vs in re.findall(r'[aeiou]{2,}', word))
fdist.most_common(10)

[('io', 549),
 ('ea', 476),
 ('ie', 331),
 ('ou', 329),
 ('ai', 261),
 ('ia', 253),
 ('ee', 217),
 ('oo', 174),
 ('ua', 109),
 ('au', 106)]

* findall은 뒤의 string에서 있는 것을 다 찾아라
* Freqdist의 variable type은 list. 
* [aeiou]{2,} = aeiou가 2개 이상 들어 있기만 하면 그것을 찾아와서 list로 만들어라. 
* 그 list에서 fdist를 받고 . most_common 하면, 상위 10개를 보여준다.
* 'io', 549 는 io가 549번 나와서 가장 자주 나왔다는 것.